src/models/pairwise_union_reach_surface.py

```diff
@@ -0,0 +1,217 @@
+          reach_curves: A list of ReachCurves to be used in model fitting and reach
+            prediction.
+          reach_points: A list of ReachPoints to which the model is to be fit. This
+            is parallel to the reach_curves list. The reach point at ith poisition
+            is drawn from ith reach curve.
+        """
+
+        self._reach_curves = copy.deepcopy(reach_curves)
+        self._n = len(reach_points)
+        self._p = len(reach_points[0].impressions)
+        super().__init__(data=reach_points, max_reach=0)
+
+    def _fit(self) -> None:
+        z, alpha = self.get_z_and_alpha()
+        self._a = self.solve_a_given_z_and_alpha(z, alpha)
+
+    def get_reach_vector(self, impressions: Iterable[int]) -> Iterable[int]:
+        """Calculates single publisher reaches for a given impression vector.
+
+        Args:
+          impressions: A list of impressions per publisher. prediction.
+
+        Returns:
+          Reach for each publisher.
+        """
+        return [
+            reach_curve.by_impressions(impression).reach()
+            for reach_curve, impression in zip(self._reach_curves, impressions)
+        ]
+
+    def by_impressions(
+        self, impressions: Iterable[int], max_frequency: int = 1
+    ) -> ReachPoint:
+
+        reach_vector = self.get_reach_vector(impressions)
+        reach_sum = sum(reach_vector)
+        overlap = sum(
+            [
+                (self._a[i * self._p + j] * reach_vector[i] * reach_vector[j])
+                / (
+                    max(
+                        self._reach_curves[i].max_reach, self._reach_curves[j].max_reach
+                    )
+                    * 2
+                )
+                for i in range(self._p)
+                for j in range(self._p)
+            ]
+        )
+
+        return ReachPoint(impressions, [reach_sum - overlap])
+
+    def get_inequality_constraints(self):
+        """Returns the inequality constraint for the Quadratic Programming solver.
+
+        3 types of inequality constraints are stacked in matrix G and vector h:
+
+          G1: a_i_j >= 0 for all i,j -> all parameters are non negative
+          G2: sum_{j=1..p} a_i_j <= 1 for all i -> row sums are no more than one
+          G3: sum_{i=1..p} a_i_j <= 1 for all j -> col sums are no more than one
+        """
+        G1 = -matrix(np.eye(self._p * self._p))
+        G2 = np.kron(np.eye(self._p), np.ones(self._p))
+        G3 = np.kron(np.ones(self._p), np.eye(self._p))
+        h = matrix(
+            np.concatenate(
+                (np.zeros(self._p * self._p), np.ones(2 * self._p)), axis=None
+            )
+        )
+        return matrix(np.vstack((G1, G2, G3))), h
+
+    def get_equality_constraints(self):
+        """Returns the equality constraint for the Quadratic Programming solver.
+
+        1 type of equality constraint is described by matrix A and vector b:
+
+          a_i_i = 0 for all i -> publishers don't have overlap with themselves.
```

Comment:
Perhaps clarify the return:
Returns (matrix A, vector b). Then, cvxopt.solvers recognizes the constraint Ax = b, where x is the parameter vector.